File: 1726/captain/Robot.py
import queue as Queue
# 동 : 1, 서 : 2, 남 : 3, 북 : 4
dx = (1, 0, -1, 0)  # 동 -> 북 -> 서 -> 남
dy = (0, -1, 0, 1)

def bfs():
    current = queue.get()
    count = current[0]
    x = current[1]
    y = current[2]
    d = current[3]
    print('\ncount : {0}, x : {1}, y : {2}, d: {3}'.format(count, x, y, d))
    if x == gx and y == gy:
        print(count if d == gd else count + 1)
        while not queue.empty():
            queue.get()
        return
    # 직진
    for i in range(1, 4):
        nx = x + dx[d] * i
        ny = y + dy[d] * i
        if -1 < nx < N and -1 < ny < M and (board[ny][nx] == 0 or board[ny][nx] > count + 1):
            board[ny][nx] = count + 1
            queue.put([count + 1, nx, ny, d])
    
    # left 회전 d + 1 
    nd = d + 1 if d < 3 else 0
    nx = x + dx[nd]
    ny = y + dy[nd]
    if -1 < nx < N and -1 < ny < M and (board[ny][nx] == 0 or board[ny][nx] > count + 2):
        queue.put([count + 1, x, y, nd])
    
    # right 회전 d - 1
    nd = d - 1 if d > 0 else 3
    nx = x + dx[nd]
    ny = y + dy[nd]
    if -1 < nx < N and -1 < ny < M and (board[ny][nx] == 0 or board[ny][nx] > count + 2):
        queue.put([count + 1, x, y, nd])


def direct(d):
    if d == 2 or d == 3:
        return d
    elif d == 1:
        return 0
    elif d == 4:
        return 1

if __name__=='__main__':
    with open('../input.txt', 'r') as f:
        M, N = map(int, f.readline().strip().split(' '))
        board = []
        for _ in range(M):
            board.append(list(map(int, f.readline().strip().split(' '))))
        start = list(map(int, f.readline().strip().split(' ')))
        goal = list(map(int, f.readline().strip().split(' ')))
        gx = goal[1] - 1
        gy = goal[0] - 1
        gd = direct(goal[2])
        board[start[0] - 1][start[1] - 1] = -1
        queue = Queue.PriorityQueue()
        queue.put([0, start[1] - 1, start[0] - 1, direct(start[2])])
        while not queue.empty():
            bfs()
            for b in board:
                print(b)

File: 1726/captain/Robot.cpp
#include <cstdio>
#include <queue>
using namespace std;

int const MAX_SIZE = 101;
int const direct[] = {0, 1, 3, 2, 0};
int const dx[] = {0, 1, 0, -1};
int const dy[] = {-1, 0, 1, 0};

struct pos {
    int x; // x 좌표
    int y; // y 좌표
    int d; // 방향
    int cnt; // 명령 횟수
};

int m, n; // 세로 길이, 가로 길이 
int board[MAX_SIZE][MAX_SIZE];
int minCount[MAX_SIZE][MAX_SIZE][4];
pos start, goal;
queue<pos> q;

int change(int d) {
    return direct[d];
}

void bfs() {
    while (!q.empty()) {
        pos cur = q.front();
        // printf("x: %d, y: %d, d: %d, cnt: %d\n", cur.x, cur.y, cur.d, cur.cnt);
        q.pop();
        int nx, ny, nd;
        // 직진
        for (int i = 1; i < 4; i++) {
            nx = cur.x + dx[cur.d] * i;
            ny = cur.y + dy[cur.d] * i;
            if (nx < 0 || nx >= n || ny < 0 || ny >= m || board[ny][nx] == 1) break;
            if (minCount[ny][nx][cur.d] == -1 || minCount[ny][nx][cur.d] > cur.cnt + 1) {
                minCount[ny][nx][cur.d] = cur.cnt + 1;
                q.push(pos({nx, ny, cur.d, cur.cnt + 1}));
            }
        }
        // 우로 회전
        nd = cur.d == 3 ? cur.d - 3 : cur.d + 1;
        if (minCount[cur.y][cur.x][nd] == -1 || minCount[cur.y][cur.x][nd] > cur.cnt + 1) {
            minCount[cur.y][cur.x][nd] = cur.cnt + 1;
            q.push(pos({cur.x, cur.y, nd, cur.cnt + 1}));
        }
        // 좌로 회전
        nd = cur.d == 0 ? cur.d + 3 : cur.d - 1;
        if (minCount[cur.y][cur.x][nd] == -1 || minCount[cur.y][cur.x][nd] > cur.cnt + 1) {
            minCount[cur.y][cur.x][nd] = cur.cnt + 1;
            q.push(pos({cur.x, cur.y, nd, cur.cnt + 1}));
        }
        // for (int i = 0; i < 5; i++) {
        //     for (int j = 0; j < 4; j++) {
        //         for (int k = 0; k < 6; k++) {
        //             printf("%3d ", minCount[i][k][j]);
        //         }
        //         printf("   ");
        //     }
        //     printf("\n");
        // }
    }
}

int main() {
    scanf("%d %d", &m, &n);
    for (int i = 0; i < m; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%d", &board[i][j]);
            for (int k = 0; k < 4; k++) {
                minCount[i][j][k] = -1;
            }
        }
    }
    scanf("%d %d %d", &start.y, &start.x, &start.d);
    scanf("%d %d %d", &goal.y, &goal.x, &goal.d);
    start.d = change(start.d);
    goal.d = change(goal.d);
    minCount[start.y - 1][start.x - 1][start.d] = 0;
    q.push(pos({start.x - 1, start.y - 1, start.d, 0}));
    bfs();
    printf("%d\n", minCount[goal.y - 1][goal.x - 1][goal.d]);
    return 0;
}